### ISSUE TITLE
gulp generic cannot create build/pdf.js and pdfwork.js

### ISSUE THREAD

**Issue by unrealyan** [mozilla/pdf.js#9737]
![image](https://user-images.githubusercontent.com/5870143/40226739-17dab798-5abf-11e8-8aa7-bad823674455.png)
![image](https://user-images.githubusercontent.com/5870143/40226745-1d5a1b46-5abf-11e8-9a99-2c56f00a2b1f.png)

can not create build/generic/build

**Comment by Rob--W** [comment]
Can't reproduce this problem on Linux (see below). Can you show the exact message that you're getting after the `viewer.js.map` line?

```sh
$ gulp generic
[12:56:02] Using gulpfile /tmp/pdf.js/gulpfile.js
[12:56:02] Starting 'buildnumber'...

### Getting extension build number
[12:56:02] Starting 'locale'...

### Building localization files
(node:17166) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
Extension build number: 502
[12:56:02] Finished 'buildnumber' after 94 ms
[12:56:02] Finished 'locale' after 173 ms
[12:56:02] Starting 'generic'...

### Creating generic viewer

### Bundling files into pdf.js
[12:56:10] Version: webpack 3.12.0
        Asset    Size  Chunks                    Chunk Names
    viewer.js  394 kB       0  [emitted]  [big]  main
viewer.js.map  632 kB       0  [emitted]         main
[12:56:17] Version: webpack 3.12.0
     Asset     Size  Chunks                    Chunk Names
    pdf.js   614 kB       0  [emitted]  [big]  main
pdf.js.map  1.01 MB       0  [emitted]         main
[12:56:19] Version: webpack 3.12.0
            Asset     Size  Chunks                    Chunk Names
    pdf.worker.js  1.52 MB       0  [emitted]  [big]  main
pdf.worker.js.map   2.8 MB       0  [emitted]         main
[12:56:19] Finished 'generic' after 17 s

```

**Comment by EvgeniyAn** [comment]
I have the same problem(Windows10):

`PS C:\Users\e.andrienko\source\repos\pdfjs\pdf.js> gulp generic
[11:35:43] Using gulpfile ~\source\repos\pdfjs\pdf.js\gulpfile.js
[11:35:43] Starting 'buildnumber'...

### Getting extension build number
[11:35:43] Starting 'locale'...

### Building localization files
Extension build number: 506
[11:35:43] Finished 'buildnumber' after 752 ms
[11:35:45] Finished 'locale' after 2.52 s
[11:35:45] Starting 'generic'...

### Creating generic viewer

### Bundling files into pdf.js
[11:36:24] Version: webpack 3.12.0
        Asset    Size  Chunks                    Chunk Names
    viewer.js  394 kB       0  [emitted]  [big]  main
viewer.js.map  679 kB       0  [emitted]         main

<--- Last few GCs --->

[13520:000001B4EFED9C30]   853983 ms: Mark-sweep 1440.5 (1478.4) -> 1440.6 (1478.4) MB, 1928.6 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0000033FD9F25EC1 <JSObject>
    1: allGeneratedPositionsFor(aka SourceMapConsumer_allGeneratedPositionsFor) [C:\Users\e.andrienko\source\repos\pdfjs\pdf.js\node_modules\babel-core\node_modules\source-map\lib\source-map-consumer.js:~178] [pc=0000039B9EE7A1B1](this=000000684A402249 <BasicSourceMapConsumer map = 000000965D4137C1>,aArgs=00000310F004B969 <Object map = 000000965D414211>)
    2: /* anonymous */(aka /* anonymous...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`

**Comment by timvandermeij** [comment]
Yes, this traceback is equal as the one we have on the bots right now. Gulp (NodeJS) is suddenly using a lot more memory and NodeJS can crash at random moments. I tried to fix this by upgrading all packages and implementing caching in the currently open build improvements PR, but no luck there yet. Thank you for reporting this because it makes it clear that it's not just a bot problem (because I cannot reproduce it locally).

**Comment by timvandermeij** [comment]
Yes, I found the same in https://github.com/mozilla/pdf.js/pull/9741#issuecomment-390487717, in which I hoped that upgrading everything would help a bit. It looks like we were close to this limit before, but recent Node/NPM/Webpack/[some other package] upgrades caused the memory usage to exceed the limit. Passing `--max-old-space-size=4096` to e.g., `node` could solve this perhaps, but I'm wondering how we can do that in a general way so that you don't have to type this for every `gulp` command. The best I could think of is to use NPM scripts in `package.json`, but that won't stop anyone from running `gulp generic` and running into this again. I tried setting the `NODE_OPTIONS="--max-old-space-size=4096"` environment variable in `gulpfile.js`, but that also didn't help (perhaps it's set too late since the process is already running?).

I'm running a bit out of options, so any help is appreciated here. I could not find the root cause of the memory usage, but I think Webpack is the culprit here since it doesn't appear to happen in other build stages. Can we limit the Webpack memory usage somehow or debug it easily?

**Comment by skyjust** [comment]
I have the same problem(Windows 7), but no problem on ubuntu 12.04

**Comment by Snuffleupagus** [comment]
> Passing --max-old-space-size=4096 to e.g., node could solve this perhaps, but I'm wondering how we can do that in a general way so that you don't have to type this for every gulp command.

There's apparently a `npm` package, see https://www.npmjs.com/package/increase-memory-limit, that claims to do this automatically. That package may be helpful, but please note that I've not tested it.
Slightly off topic, but it seems ridiculous that you'd need close to 4 GB of memory just to successfully build a JavaScript library!

**Comment by timvandermeij** [comment]
Yes, indeed. That's why I'm not happy at all with that solution in the long term. For the short term it could help to make the bots usable again, but there is definitely a root cause that I just can't figure out. The memory usage is far too high.

**Comment by pravid** [comment]
Any solution to this yet?
I'm getting the same issue while building the project. I tried the solution suggested by, Snuffleupagus but it's not working.
I'm using, Windows10, 32 bit, 4GB RAM
Here is the log
```
$ gulp generic
[17:41:13] Using gulpfile d:\Projects\pdfjs\pdf.js\gulpfile.js
[17:41:13] Starting 'buildnumber'...

### Getting extension build number
[17:41:13] Starting 'locale'...

### Building localization files
Extension build number: 511
[17:41:14] Finished 'buildnumber' after 982 ms
[17:41:14] Finished 'locale' after 1.11 s
[17:41:14] Starting 'generic'...

### Creating generic viewer

### Bundling files into pdf.js
[17:41:42] Version: webpack 3.12.0
        Asset    Size  Chunks                    Chunk Names
    viewer.js  394 kB       0  [emitted]  [big]  main
viewer.js.map  679 kB       0  [emitted]         main

<--- Last few GCs --->

[3196:00409018]   438955 ms: Mark-sweep 700.8 (743.7) -> 700.8 (743.7) MB, 1548.
7 / 0.0 ms  allocation failure GC in old space requested
[3196:00409018]   440393 ms: Mark-sweep 700.8 (743.7) -> 700.7 (729.7) MB, 1438.
2 / 0.0 ms  last resort GC in old space requested
[3196:00409018]   441841 ms: Mark-sweep 700.7 (729.7) -> 700.8 (729.7) MB, 1448.
1 / 0.0 ms  last resort GC in old space requested


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0361632D <JSObject>
    1: /* anonymous */(aka /* anonymous */) [d:\Projects\pdfjs\pdf.j
s\node_modules\babel-core\lib\transformation\file\merge-map.js:~296] [pc=0A5FA41
E](this=034041A1 <undefined>,item=0C0ECC75 <Object map = 060F09F9>)
    2: arguments adaptor frame: 3->1
    3: map(this=0C0F0241 <JSArray[8644]>)
    4: /* anonymous */(aka /* anonymous */) [d:\Projects\pdfjs\pdf.j
s\node_modules\ba...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node_module_register
 2: v8::internal::Factory::NewFillerObject
 3: v8::internal::operator<<
```

**Comment by brendandahl** [comment]
This should be fixed over in #9759
